src/verifier/verifier.rs

```diff
@@ -19,7 +27,535 @@ pub fn build_symbol_table(ast: &mut Whamm, err: &mut ErrorGen) -> SymbolTable {
+            } => {
+                let cond_ty = self.visit_expr(cond);
+                if let Some(ty) = cond_ty {
+                    if ty != DataType::Boolean {
+                        self.err.type_check_error(
+                            false,
+                            "Condition must be of type boolean".to_owned(),
+                            &<Option<crate::parser::types::Location> as Clone>::clone(loc)
+                                .map(|l| l.line_col),
+                        );
+                    }
+                }
+                // TODO: should conseq and alt have the same type?
+                let _conseq_ty = self.visit_expr(conseq);
+                let _alt_ty = self.visit_expr(alt);
+                todo!()
+            }
+        }
+    }
+
+    #[allow(unused_variables)]
+    fn visit_datatype(&mut self, datatype: &mut DataType) -> Option<DataType> {
+        unimplemented!()
+    }
+
+    #[allow(unused_variables)]
+    fn visit_value(&mut self, val: &mut Value) -> Option<DataType> {
+        match val {
+            Value::Integer { .. } => Some(DataType::I32),
+            Value::Str { .. } => Some(DataType::Str),
+            Value::Boolean { .. } => Some(DataType::Boolean),
+            Value::Tuple { ty, vals } => {
+                // Alex TODO: this ty does not contain the DataType
+                // and I need to recurse to get the type, why?
+                let tys = vals
+                    .iter_mut()
+                    .map(|val| self.visit_expr(val))
+                    .collect::<Vec<_>>();
+                // assume these expressions (actually just values) all parse
+                // and have Some type
+                let mut all_tys: Vec<Box<DataType>> = Vec::new();
+                for ty in tys {
+                    if let Some(ty) = ty {
+                        all_tys.push(Box::new(ty));
+                    } else {
+                        self.err
+                            .unexpected_error(true, Some(UNEXPECTED_ERR_MSG.to_string()), None);
+                    }
+                }
+                Some(DataType::Tuple {
+                    ty_info: Some(all_tys),
+                })
+            }
+        }
+    }
+
+    #[allow(unused_variables)]
+    fn visit_formal_param(&mut self, param: &mut (Expr, DataType)) -> Option<DataType> {
+        unimplemented!()
+    }
+
+    #[allow(unused_variables)]
+    fn visit_binop(&mut self, binop: &mut BinOp) -> Option<DataType> {
+        unimplemented!()
+    }
+
+    #[allow(unused_variables)]
+    fn visit_unop(&mut self, unop: &mut UnOp) -> Option<DataType> {
+        unimplemented!()
+    }
+}
+
+pub fn type_check(ast: &mut Whamm, st: &SymbolTable, err: &mut ErrorGen) -> bool {
+    // Alex TODO, is cloning the way to go ??
+    let mut type_checker = TypeChecker {
+        table: st.clone(),
+        err: err.clone(),
+    };
+    type_checker.visit_whamm(ast);
+    // propagate error
+    *err = type_checker.err;
```

Comment:
Just pass a reference to the errorgen instead.

Reply:
Not sure about this part!